Merge Mip 203

rdenarie opened the pull request:
This commit add the portlet Login Form

<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->


<!-- Tips : 
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

rdenarie committed:
feat: Create page login in global site - MEED-9372 - meeds-io/MIPs#203

This fix create a new page login in global site, so that the page is editable inline
It add feature to make it reachable on /portal/login, and execute loginHandler if necessary

In addition, to create the layout of the page with xml, it fix a problem in binding.xml file, which have a duplicated line, preventing to set text-title-color correctly.

rdenarie committed:
feat: Add portlet Login Form - meeds-io/MIPs#203 - MEED-9360

This commit add the portlet Login Form

rdenarie committed:
feat : Make External Onboarding page modifiable in layout - MEED-9700 - meeds-io/MIPs#203

This commit update External Onboarding page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Forgot password page modifiable in layout - MEED-9362 - Meeds-io/MIPs#203

This commit update Forgot password page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Internal Onboarding page modifiable in layout - MEED-9711 - Meeds-io/MIPs#203
This commit update Internal Onboarding page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Register page modifiable in layout - MEED-9361 - Meeds-io/MIPs#203

This commit update Register page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
fix: Clean up code after feature completion - MEED-9753 - meeds-io/MIPs#203

Remove uneeded line in handlers

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/LoginHandler.java:
```diff
@@ -218,6 +218,19 @@ public String getContextPath() {
+            // with the error message
+            try {
+              String redirectUrl = referer.contains("?") ? "&" : "?";
+              redirectUrl = referer+redirectUrl+"error=" + status.getErrorCode();
```
```suggestion
              redirectUrl = "%s%serror=%s".formatted(referer, redirectUrl, status.getErrorCode());
```

rdenarie replied:
done

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/LoginHandler.java:
```diff
@@ -218,6 +218,19 @@ public String getContextPath() {
           LOG.debug("User {} authentication failed.", username, e);
           status = LoginStatus.FAILED;
+          String referer = request.getHeader("Referer");
+          if (referer!=null && !referer.contains("/portal/login?")) {
```
Code format


```suggestion
          if (referer != null && !referer.contains("/portal/login?")) {
```

rdenarie replied:
done

boubaker commented on component/web/security/src/main/java/org/exoplatform/web/login/onboarding/OnboardingHandler.java:
```diff
@@ -305,4 +285,14 @@ private void writeImage(HttpServletResponse response, BufferedImage bi) {
   }
+
+  private static void returnSubmissionError(String message, HttpServletResponse response) throws
+                                                                                                        IOException {
```
Code format

rdenarie replied:
done

boubaker commented on webui/src/main/java/org/exoplatform/portal/application/PageRequestHandler.java:
```diff
@@ -0,0 +1,344 @@
+import java.util.ServiceLoader;
+
+@SuppressWarnings("rawtypes")
+public class PageRequestHandler extends WebRequestHandler {
```
Since it's almost the same behavior than `PortalRequestHandler`, would it better to handle standalone pages in `PortalRequestHandler` rather than duplicating the code ? WDYT ?

rdenarie replied:
When I tried to use the same handler, I didnt manage to make it work correctly. 

In fact, If I use only PortalResquestHandler, when I access to /portal/login this code 
`String requestSiteName = controllerContext.getParameter(REQUEST_SITE_NAME);
String requestSiteType = controllerContext.getParameter(REQUEST_SITE_TYPE);`

set requestSiteName to "login", and requestSiteType to "portal"
Then, it search a portal named "login", which not exists.

By having firstly PageRequestHandler, which search the page "login" in "global" portal, AND if not found we call PortalRequestHandler, we ensure to find the page before.

What I can try is to make PageRequestHandler extension of PortaltRequestHandler, to only override modified method, but I think we cannot make all the stuff in one handler

boubaker replied:
You can make a list of extensible `paths`, which will make the handler handles the page as it's part of `global` site.
If the request URI is one of those paths, you will be able to force the Site Type and Name as made in `PageRequestHandler`. This list of paths can be provided in controller.xml (as param) or using plugin definitions, which makes the list easily extensible by addon (Could redefine the Download Page of documents by example to be an editable page rather than a JSP).

boubaker replied:
We may even imagine something more dynamic. All page requests not having a valid Site will be considered as a request to display a global site navigation, WDYT ?

rdenarie replied:
As I understand, you mean to list in controller.xml all pages which are in global site which should be accessible on /portal/xxx
Is it right  ?

The current solution offers the advantage to work for all pages : 
when accessing to /portal/pageName, if the page is present in global navigation, it works, without having to declare it in  addon. 
Example page : /portal/download-document for public document download, on exoplatform side

rdenarie replied:
> We may even imagine something more dynamic. All page requests not having a valid Site will be considered as a request to display a global site navigation, WDYT ?

how do you define "valid site" ? 
Should I call processRequest(context, app); firstly, and if nothing found, recall it with "global" sitename ?

boubaker replied:
nope, before `processRequest` (see attachment)

<img width="1651" height="650" alt="image" src="https://github.com/user-attachments/assets/ed9cb4cd-0f4c-4e88-9383-6318182fa14a" />

rdenarie replied:
Done.
I had to add a specific boolean globalPageAccess

This is used to prevent the access a page through a url like /portal/global/pageName to keep current behaviour

boubaker replied:
If we let accessing global pages, from now on using `/portal/{nav_name}`, we should allow to use `/portal/global/{nav_name}`. WDYT ?

rdenarie replied:
No pb for me.

@srenault-meeds are you ok with this ?

srenault-meeds replied:
ok for me too

rdenarie replied:
Done

rdenarie committed:
fix : Correct feedbacks for merge - meeds-io/MIPs#203 - MEED-9791

- Remove PageRequestHandler and make all stuff in PortalRequestHandler
- Finalize ForgotPassword transformation from handler to portlet
- Finalize Register transformation from handler to portlet
- Finalize External Onboarding transformation from handler to portlet
- Finalize Internal Onboarding transformation from handler to portlet
- This commit reduce the code coverage ratio. As we removed handlers which have a hight coverage ratio, the average coverage reduces.
- Change a string "RegisterExtension" used few times to a constant
